Add read-only Task endpoint

turbaszek opened the pull request:
Closes #8138 

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

Commit:
Add read-only Task endpoint

fixup! Add read-only Task endpoint

Add tests

Add tests for serialized dag

Fix tests

Fix requirements

Refactor task endpoint

mik-laj commented on airflow/api_connexion/schemas/common_schema.py:
```diff
@@ -0,0 +1,145 @@
+            raise Exception("Unknown object type: {}".format(obj.__class__.__name__))
+
+
+ColorField = partial(fields.String, validate=validate.Regexp("^#[a-fA-F0-9]{3,6}$"))
```
Could you do less magic?
https://github.com/apache/airflow/blob/75bd2f180e4afb591b08ee5d2cdb9b85737258b8/airflow/api_connexion/schemas/enum_schemas.py

mik-laj commented on setup.py:
```diff
@@ -708,6 +708,8 @@ def is_package_excluded(package: str, exclusion_list: List[str]):
     'lazy_object_proxy~=1.3',
     'lockfile>=0.12.2',
     'markdown>=2.5.2, <3.0',
+    'marshmallow<3',
```
We don't suppport marsshmalllow yet. We have a dependency on FAB, which requires marshmallow 2.0 Support for 3.0 is WIP: https://github.com/dpgaspar/Flask-AppBuilder/pull/1382

mik-laj commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -1337,7 +1335,6 @@ components:
         extra:
           type: string
           readOnly: true
-          nullable: true
```
?

turbaszek replied:
Mistake, fixed

turbaszek committed:
fixup! Add read-only Task endpoint

turbaszek committed:
Add DAG details endpoint

fixup! Add DAG details endpoint

fixup! fixup! Add DAG details endpoint

fixup! fixup! fixup! Add DAG details endpoint

fixup! fixup! fixup! fixup! Add DAG details endpoint

turbaszek commented on airflow/api_connexion/schemas/dag_schema.py:
```diff
@@ -0,0 +1,93 @@
+    fileloc = auto_field(dump_only=True)
+    owners = fields.Method("get_owners", dump_only=True)
+    description = auto_field(dump_only=True)
+    # schedule_interval = fields.Nested(ScheduleIntervalSchema, dump_only=True)
```
Had to figure what's wrong:
```
FAILED [100%][2020-06-19 16:27:55,965] {validation.py:230} ERROR - http://localhost/api/v1/dags/test_dag/details validation error: {'__type': 'TimeDelta', 'days': 1, 'seconds': 0} is valid under each of {'x-scope': ['', '#/components/schemas/DAGCollection', '#/components/schemas/DAG', '#/components/schemas/ScheduleInterval'], 'type': 'object', 'required': ['__type'], 'properties': {'__type': {'type': 'string'}, 'years': {'type': 'integer'}, 'months': {'type': 'integer'}, 'days': {'type': 'integer'}, 'leapdays': {'type': 'integer'}, 'hours': {'type': 'integer'}, 'minutes': {'type': 'integer'}, 'seconds': {'type': 'integer'}, 'microseconds': {'type': 'integer'}, 'year': {'type': 'integer'}, 'month': {'type': 'integer'}, 'day': {'type': 'integer'}, 'hour': {'type': 'integer'}, 'minute': {'type': 'integer'}, 'second': {'type': 'integer'}, 'microsecond': {'type': 'integer'}}}, {'x-scope': ['', '#/components/schemas/DAGCollection', '#/components/schemas/DAG', '#/components/schemas/ScheduleInterval'], 'type': 'object', 'required': ['__type'], 'properties': {'__type': {'type': 'string'}, 'value': {'type': 'string'}}}, {'x-scope': ['', '#/components/schemas/DAGCollection', '#/components/schemas/DAG', '#/components/schemas/ScheduleInterval'], 'type': 'object', 'required': ['__type'], 'properties': {'__type': {'type': 'string'}, 'days': {'type': 'integer'}, 'seconds': {'type': 'integer'}, 'microsecond': {'type': 'integer'}}}
Failed validating 'oneOf' in schema['allOf'][0]['properties']['schedule_interval']:
    {'discriminator': {'propertyName': '__type'},
     'oneOf': [{'properties': {'__type': {'type': 'string'},
                               'days': {'type': 'integer'},
                               'microsecond': {'type': 'integer'},
                               'seconds': {'type': 'integer'}},
                'required': ['__type'],
                'type': 'object',
                'x-scope': ['',
                            '#/components/schemas/DAGCollection',
                            '#/components/schemas/DAG',
                            '#/components/schemas/ScheduleInterval']},
               {'properties': {'__type': {'type': 'string'},
                               'day': {'type': 'integer'},
                               'days': {'type': 'integer'},
                               'hour': {'type': 'integer'},
                               'hours': {'type': 'integer'},
                               'leapdays': {'type': 'integer'},
                               'microsecond': {'type': 'integer'},
                               'microseconds': {'type': 'integer'},
                               'minute': {'type': 'integer'},
                               'minutes': {'type': 'integer'},
                               'month': {'type': 'integer'},
                               'months': {'type': 'integer'},
                               'second': {'type': 'integer'},
                               'seconds': {'type': 'integer'},
                               'year': {'type': 'integer'},
                               'years': {'type': 'integer'}},
                'required': ['__type'],
                'type': 'object',
                'x-scope': ['',
                            '#/components/schemas/DAGCollection',
                            '#/components/schemas/DAG',
                            '#/components/schemas/ScheduleInterval']},
               {'properties': {'__type': {'type': 'string'},
                               'value': {'type': 'string'}},
                'required': ['__type'],
                'type': 'object',
                'x-scope': ['',
                            '#/components/schemas/DAGCollection',
                            '#/components/schemas/DAG',
                            '#/components/schemas/ScheduleInterval']}],
     'readOnly': True,
     'x-scope': ['',
                 '#/components/schemas/DAGCollection',
                 '#/components/schemas/DAG']}
On instance['schedule_interval']:
    {'__type': 'TimeDelta', 'days': 1, 'seconds': 0}
```

turbaszek replied:
The problem seems to be solved  by adding more than `__type` to `required` fields.

turbaszek commented on tests/api_connexion/endpoints/test_dag_endpoint.py:
```diff
@@ -14,35 +14,109 @@
     def setUp(self) -> None:
+        self.clean_db()
         self.client = self.app.test_client()  # type:ignore
+        self.client_serialized = self.app_serialized.test_client()  # type:ignore
```
I am not sure if this is the right way but nothing else comes to my mind. What we have here are two testing clients one with single DAG dagbag and second one with empty one with serialization switched on

mik-laj replied:
Maybe just give up this base class and divide the whole into two classes?

turbaszek replied:
I was thinking about creating a new app in test with serialization, WDYT?

ephraimbuddy commented on airflow/api_connexion/endpoints/task_endpoint.py:
```diff
@@ -14,20 +14,36 @@
+    if not dag:
+        raise NotFound("DAG not found")
+    tasks = dag.tasks
+    task_collection = TaskCollection(tasks=tasks, total_entries=len(tasks))
```
The` total_entries` is not actually the total of query result. In this case now, It is the total of all tasks in the db.

turbaszek replied:
Hm, I think here we return only tasks from single dag, so if I correctly understand the `total_entries` should be equal to the number of tasks in the requested DAG

mik-laj replied:
This endpoint has no pagination, because all objects must be loaded to be able to return a response. We don't use the database here, so we can't optimize it.

mik-laj commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -1964,13 +1972,14 @@ components:
       type: object
       required:
         - __type
+        - years
```
Should we add other fields also?

turbaszek replied:
Done

mik-laj commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -1964,13 +1972,14 @@ components:
         __type: {type: string}
```
Should add other fields to required also?

turbaszek replied:
Done

mik-laj commented on tests/cli/commands/test_dag_command.py:
```diff
@@ -59,8 +60,14 @@ class TestCliDags(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
         cls.dagbag = DagBag(include_examples=True)
+        DAG.bulk_sync_to_db([d[1] for d in cls.dagbag.dags.items()])
```
```suggestion
        DAG.bulk_sync_to_db([d for d in cls.dagbag.dags.values()])
```
or
```suggestion
       cls.dagbag.sync_to_db()
```

turbaszek replied:
Done, thanks!

turbaszek committed:
fixup! Add DAG details endpoint

turbaszek committed:
fixup! fixup! Add DAG details endpoint

turbaszek committed:
Update tests
